### ISSUE TITLE
[RFC] Extension Script Loading Additions 

### ISSUE THREAD

**Issue by cdowdy** [bolt/bolt#2736]
I was curious if you all would be willing to add one more option to the script loading behavior for Extensions which would be the option to load the scripts with ```asnyc defer```.  

Currently there are 2 options you can pass: ```late``` and ```priority``` -besides filename.   

If I understand the docs ```late``` simply determines if the script should be loaded in the ```<head```> and ```priority``` determines source order(?).  I'm proposing another option ```$async = false``` (as a default to devs who don't care about blocking rendering) to [BaseExtension.php](https://github.com/bolt/bolt/blob/master/src/BaseExtension.php#L423-L437)   
```php  
public function addJavascript($filename, $late = false, $priority = 0, $async = false)
    {
        // check if the file exists.
        if (file_exists($this->basepath . "/" . $filename)) {
            // file is located relative to the current extension.
            $this->app['extensions']->addJavascript($this->getBaseUrl() . $filename, $late, $priority, $async);
        } elseif (file_exists($this->app['paths']['themepath'] . "/" . $filename)) {
            // file is located relative to the theme path.
            $this->app['extensions']->addJavascript($this->app['paths']['theme'] . $filename, $late, $priority, $async);
        } else {
            // Nope, can't add the CSS.
            $message = "Couldn't add Javascript '$filename': File does not exist in '" . $this->getBaseUrl() . "'.";
            $this->app['logger.system']->addError($message, array('event' => 'extensions'));
        }
    }  
```  

Then in the [processAssets](https://github.com/bolt/bolt/blob/master/src/Extensions.php#L629-L666) function we would add ```async defer``` and the resulting script would then be rendered as   
```html  
<script src="filename.js" async defer></script>  
```  
Browser support for ```async``` is "modern browsers" ( Internet Explorer 10+ Android 3.0+) adding ```defer``` gives you support back to Internet Explorer 5.5 and for other browsers who don't recognize it will be treated as a regular rendering blocking js file.    

Maybe instead of passing a boolean we could pass ```$async``` as a string to give developers even more control on whether to just give ```async```, ```defer``` or both as ```async defer```  

As the extension library grows this could keep us from having to do hacks like many of us have done for older (< 4.1) WordPress sites to keep a page from having rendering issues. In my opinion this could/should be a core behavior instead of another extension.  

For reasons on why to use async defer together you can see Ilya Grigorik's [Script-injected "async scripts" considered harmful](https://www.igvita.com/2014/05/20/script-injected-async-scripts-considered-harmful/)

Any concerns about this? Additions? Or not interested?

**Comment by GawainLynch** [comment]
:+1: but me and frontend... :grin:

**Comment by cdowdy** [comment]
> but me and frontend

yeah this would only ideally be used in extensions that deal with front end page rendering. Dealing with script loading and their order in the backend is little more complicated.  

My thinking is for extensions like social share buttons and the like. Bolt already is extremely lightweight and doesn't make any assumptions or decisions for you (!!!!) when it comes to what you want your site to look like. The last thing __I__ want is to spend time with webpagetest/sitespeed.io/pagespeed insights, work on a good score only to fight an extension that loads its JS in the head, in a blocking manner or its css underneath a js file/in the footer. Having this option there for a dev to say "hey this wont hurt page render lets load it async" or for a person who is comfortable changing the code itself to not block page render would be nice ( and keep with the dev friendly theme you all have already created).

**Comment by GawainLynch** [comment]
> The last thing I want is to spend time with webpagetest/sitespeed.io/pagespeed insights

Well I would say that is precisely why I run ngx_pagespeed...  That doesn't invalidate your point, but would explain why I haven't pursued this one myself up until now. :blush:

**Comment by cdowdy** [comment]
yeah I used nginx pagespeed for a while then just used moved to serving webp with nginx itself, zopfli for serving pre compressed assets and fastcgi cache... I might go back just to use their easy localstorage haha

**Comment by GawainLynch** [comment]
Also have FastCGI caches but have them turned off from a bug that I can't quite track down...  But yeah, the Pagespeed stuff is nice when you have it down.

**Comment by bobdenotter** [comment]
I like this idea, but i'd like it more if the 'options' were an array: 

```
public function addJavascript($filename, $options)
{ .. }

$this->addJavascript("foo.js", array('late' => false, 'priority' => 0, 'async' => false));
```

By testing the type of the parameter we can keep this Backwards compatible, and allow for future options as well.

**Comment by SahAssar** [comment]
Seems fixed by #3078 to me. Feel free to reopen if that's not the case.
